File: dotnet/Allmark/Inline/HighlightRule.cs
namespace Allmark.Inline;

using Allmark.Types;

public static class HighlightRule
{
    public static InlineRule Create()
    {
        return new InlineRule
        {
            Name = "highlight",
            Test = TestHighlight,
            Precedence = 5,
        };
    }

    private static bool TestHighlight(InlineParserState state, MarkdownNode parent)
    {
        var ch = Utils.GetChar(state.Src, state.I);
        if (!state.IsEscaped && ch == '=')
        {
            var rule = Create();
            return TagMarksRule.Execute("highlight", ch.ToString(), state, parent, rule.Precedence!.Value);
        }
        return false;
    }
}

File: web/src/inline/highlightRule.ts
import type InlineParserState from "../types/InlineParserState";
import type InlineRule from "../types/InlineRule";
import type MarkdownNode from "../types/MarkdownNode";
import { EQUALS_CODE } from "../utils/charCodes";
import testTagMarks from "./tagMarksRule";

const rule: InlineRule = {
	name: "highlight",
	test: testHighlight,
	precedence: 5,
};
export default rule;

function testHighlight(state: InlineParserState, parent: MarkdownNode): boolean {
	if (!state.isEscaped && state.src.charCodeAt(state.i) === EQUALS_CODE) {
		return testTagMarks(rule.name, "=", state, parent, rule.precedence!);
	}
	return false;
}
